util/patch/options.go

```diff
@@ -37,6 +37,14 @@ type HelperOptions struct {
 	// OwnedConditions defines condition types owned by the controller.
 	// In case of conflicts for the owned conditions, the patch helper will always use the value provided by the controller.
 	OwnedConditions []clusterv1.ConditionType
+
+	// PatchLogging enables the patch helper to log the content of the patch.
+	// Warning: This will log all keys and values from a patch including any sensitive information contained within.
+	// To sanitize information in the patch, paths to be redacted can be supplied to WithSanitizedLoggingPaths.
+	PatchLogging bool
```

Comment:
Instead of a boolean flag, have we considered using a verbosity threshold?

Reply:
The idea is that we never under any circumstances log secrets. With a verbosity threshold we would log secrets if the log level is high enough.

Reply:
Can't we detect secret types and refuse to log diffs in that case? The paths based redaction of critical information seems a bit flaky given that APIs change. Should we rather try to find a different approach by annotating fields instead and building an internal map from GVKs?

Reply:
Kubernetes initially considered this, but they dropped that feature again: https://github.com/kubernetes-sigs/cluster-api/issues/3889#issuecomment-1043166551. IIRC due to runtime performance concerns.
They kept the datapolicy tags for static linting however (I also mentioned that on the linked issue).

I think it would be relatively hard to maintain a library ourselves which goes over structs and matches on datapolicy tags. I'm not sure if it's even possible given that we're working on unstructured most of the time and don't even have access to the structs.